File: 2024/8/30/programmers1.py
import heapq


def solution(n, works):
    answer = 0
    hq = []
    for v in works:
        heapq.heappush(hq, -v)
    while (n > 0):
        v = heapq.heappop(hq)
        if v == 0: break
        v += 1
        heapq.heappush(hq, v)
        n -= 1

    for v in hq:
        answer += (v * -1) ** 2
    return answer

File: 2024/8/30/programmers1.cpp
#include <string>
#include <vector>
#include <queue>
#include <cmath>
using namespace std;

long long solution(int n, vector<int> works) {
    long long answer = 0;
    // priority_queue<int, vector<int>, greater<int>> pq;
    priority_queue<int> pq;

    for (const int& work : works) pq.push(work);

    while (n > 0){
        int v = pq.top();
        pq.pop();
        if (v < 1) break;
        v -= 1;
        pq.push(v);
        n -= 1;
    }
    while (!pq.empty()){
        int v = pq.top();
        pq.pop();
        answer += pow(v, 2);
    }

    return answer;
}